pkg/volume/csi/csi_attacher.go

```diff
@@ -197,8 +197,18 @@ func (c *csiAttacher) VolumesAreAttached(specs []*volume.Spec, nodeName types.No
 		}
 
 		attachID := getAttachmentName(volumeHandle, driverName, string(nodeName))
+		var attach *storage.VolumeAttachment
+		if c.plugin.volumeAttachmentLister != nil {
+			attach, err = c.plugin.volumeAttachmentLister.Get(attachID)
+			if err == nil {
+				attached[spec] = attach.Status.Attached
```

Comment:
From the discussion in https://github.com/kubernetes-csi/external-attacher/issues/215#issuecomment-601065201, we should only use the absence of the VolumeAttachment object to indicate that the volume is detached. If the `status.attached` is false, that could mean that we're in the process of attaching, and we probably don't want to mark it as detached here.  Same with L218.

Do we want to make the change in this PR or separately? 

Also can potentially be investigated separately, but is it possible to have a race where we're starting to Attach, but haven't created the VolumeAttachment object yet? Then this comes along and marks it as detached? AFIAK, the Attach operartion is launched as a separate go routine and we don't wait for it to complete before coming here.

cc @jsafrane

Reply:
I think A/D controller calls `VolumesAreAttached` for volumes that are marked as attached. So in the worst case, A/D controller starts detach + `VolumesAreAttached` in parallel and `VolumesAreAttached` can mark the volume as detached earlier than `Detach()` would do.

Is it possible to attach, VolumesAreAttached, detach, attach a volume so quickly that the second attach succeeds before VolumesAreAttached evaluates the old attachment and considers it detached? As consequence, Attach would be called again and it succeeds immediately.